snakemake_executor_plugin_interface/__init__.py

```diff
@@ -57,51 +124,52 @@ def __init__(self, executor_base_cls: type[object]):
+        # They should follow the same convention if on pip, snakemake-executor-<name>
+        # Note that these will not be detected installed in editable mode (pip install -e .)
+        for _, name, _ in pkgutil.iter_modules():
+            if (
+                not name.startswith(common.executor_plugin_module_prefix)
+                or name == "snakemake_executor_plugin_interface"
+            ):
+                continue
+            module = importlib.import_module(name)
+            self._register_plugin(name, module)
 
     def _register_plugin(self, name: str, plugin: types.ModuleType):
+        """Validate and register a plugin"""
         self._validate_plugin(name, plugin)
-        self.plugins[name] = Plugin(
-            name,
+
+        # Derive the shortened name for future access
+        plugin_name = name.replace(common.executor_plugin_module_prefix, "")
+
+        self.plugins[plugin_name] = Plugin(
+            plugin_name,
             plugin.Executor,
             common_settings=getattr(plugin, "common_settings", None),
             _executor_settings_cls=getattr(plugin, "ExecutorSettings", None),
         )
 
     def _validate_plugin(self, name: str, module: types.ModuleType):
-
+        """Validate a plugin for attributes and naming"""
         expected_attributes = {
-           "common_settings": Optional[CommonSettings],
-           "ExecutorSettings": Optional[type[ExecutorSettingsBase]],
-           "Executor": type[self.executor_base_cls],
+            "common_settings": Optional[CommonSettings],
+            "ExecutorSettings": Optional[type[ExecutorSettingsBase]],
+            "Executor": type[self.executor_base_cls],
         }
-
         for attr, attr_type in expected_attributes.items():
             # check if attr is missing and fail if it is not optional
             is_optional = type(attr_type) == typing._UnionGenericAlias
             if not hasattr(module, attr):
                 if is_optional:
                     continue
-                else:
-                    raise InvalidPluginException(
-                        name, f"plugin does not define {attr}."
-                    )
-            if is_optional:
-                # get inner type
-                attr_type, = attr_type.__args__
-            attr_value = getattr(module, attr)
-            if type(attr_type) == types.GenericAlias:
-                # check for class type
-                cls, = attr_type.__args__
-                if not issubclass(attr_value, cls):
-                    raise InvalidPluginException(name, f"{attr} must be a subclass of {cls.__module__}.{cls.__name__}.")
-            else:
-                # check for instance type
-                if not isinstance(attr_value, attr_type):
-                    raise InvalidPluginException(
-                        name, f"{attr} must be of type {attr_type}."
-                    )
```

Comment:
Why this change?

Reply:
This errored and I didn’t understand what you were trying to do enough to debug it.

Reply:
> I was confused about what you were doing in the validation function, so I removed the part I didn't understand for now to get it working. I also don't understand why you require an executor_base_class to init the registry (e.g., Plugin) because it was checking for the direct class and would also need to check for parent classes to validate that. It also wouldn't make sense to instantiate that with any single Plugin class (and not the base class) because the registry is supposed to handle multiple.

I'll answer here. This code is supposed to do the following. At the beginning, we define attributes that the plugin should have (optional common setttings, optional executor settings, and (mandatory) a class called "Executor" that inherits from the AbstractExectutor class of Snakemake (which we pass to the registry as an argument coming from snakemake because we don't want to depend on snakemake from the interface package to avoid a circular dependency).

For each of these things, we check whether they are present with the right type here. The optional ones may be not present at all. I tried to implement this using the typing api, but I haven't checked whether the code is runnable yet, so there might be of course some bugs. But let us keep the code and fix it instead of deleting it.

Reply:
Working on the tests now, I will reinsert the code and see what I did wrong.